pkg/lib/scoped/finder.go

```diff
@@ -0,0 +1,93 @@
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NewUserDefinedServiceAccountFinder returns a new instance of OperatorGroupFinder.
+func NewUserDefinedServiceAccountFinder(logger *logrus.Logger, crclient versioned.Interface) *UserDefinedServiceAccountFinder {
+	return &UserDefinedServiceAccountFinder{
+		logger:   logger,
+		crclient: crclient,
+	}
+}
+
+// UserDefinedServiceAccountFinder retrieves reference to user defined service account(s).
+type UserDefinedServiceAccountFinder struct {
+	crclient versioned.Interface
+	logger   *logrus.Logger
+}
+
+// NamespaceGetter returns an instance of ServiceAccountGetterFunc that can be used by the
+// caller to get the reference to the service account associated with the namespace.
+func (f *UserDefinedServiceAccountFinder) NamespaceGetter(namespace string) ServiceAccountGetterFunc {
+	getter := func() (reference *corev1.ObjectReference, err error) {
+		logger := f.logger.WithFields(logrus.Fields{
+			"namespace":  namespace,
+			logFieldName: logFieldValue,
+		})
+
+		return GetServiceAccountFromNamespace(logger, f.crclient, namespace)
+	}
+
+	return getter
+}
+
+// GetServiceAccountFromNamespace will return the reference to a service account
+// associated with the operator group for the given namespace.
+// - If no operator group is found in the namespace, both reference and err are set to nil.
+// - If an operator group found is not managing the namespace then it is ignored.
+// - If no operator group is managing this namespace then both reference and err are set to nil.
+// - If more than one operator group are managing this namespace then an error is thrown.
+func GetServiceAccountFromNamespace(logger *logrus.Entry, crclient versioned.Interface, namespace string) (reference *corev1.ObjectReference, err error) {
+	// TODO: use a lister instead of a noncached client here.
+	list, err := crclient.OperatorsV1().OperatorGroups(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return
+	}
+
+	if len(list.Items) == 0 {
+		logger.Warnf("list query returned an empty list")
+		return
+	}
+
+	groups := make([]*v1.OperatorGroup, 0)
+	for _, og := range list.Items {
+		if len(og.Status.Namespaces) == 0 {
+			logger.Warnf("skipping operator group since it is not managing any namespace og=%s", og.GetName())
+			continue
+		}
+
+		groups = append(groups, &og)
+	}
+
+	if len(groups) == 0 {
+		logger.Warn("no operator group found that is managing this namespace")
+		return
+	}
+
+	if len(groups) > 1 {
+		err = fmt.Errorf("more than one operator group(s) are managing this namespace count=%d", len(groups))
+		return
+	}
+
+	group := groups[0]
+	if group.Spec.ServiceAccount == "" {
+		// No user defined service account is specified.
+		return
+	}
+
+	if group.Status.ServiceAccountRef == nil {
```

Comment:
If `ServiceAccountRef == nil` but `ServiceAccount != ""`, that means the user specified a serviceaccount but it couldn't be found. I think we should probably error out in that case with a message for the user.

Reply:
Does it sound better? - `Please ensure that the service account {name} exists`